src/backend/distributed/transaction/remote_transaction.c

```diff
@@ -856,22 +857,44 @@ CheckTransactionHealth(void)
 
 
 /*
- * Assign2PCIdentifier compute the 2PC transaction name to use for a
- * transaction.
+ * Assign2PCIdentifier computes the 2PC transaction name to use for a
+ * transaction. Every prepared transaction should get a new name, i.e. this
+ * function will need to be called again.
  *
- * Every 2PC transaction should get a new name, i.e. this function will need
- * to be called again.
+ * The format of the name is:
+ *
+ * citus_<source group>_<pid>_<distributed transaction number>_<connection number>
+ *
+ * (at most 5+1+10+1+10+20+1+10 = 58 characters, while limit is 64)
+ *
+ * The source group is used to distinguish 2PCs started by different
+ * coordinators. A coordinator will only attempt to recover its own 2PCs.
+ *
+ * The pid is used to distinguish different processes on the coordinator, mainly
+ * to provide some entropy across restarts.
+ *
+ * The distributed transaction number is used to distinguish different
+ * transactions originating from the same node (since restart).
+ *
+ * The connection number is used to distinguish connections made to a node
+ * within the same transaction.
  *
  * NB: we rely on the fact that we don't need to do full escaping on the names
  * generated here.
  */
 static void
 Assign2PCIdentifier(MultiConnection *connection)
 {
-	static uint64 sequence = 0;
+	/* local sequence number used to distinguish different connections */
+	static uint32 connectionNumber = 0;
+
+	/* transaction identifier that is unique across processes */
+	DistributedTransactionId *distributedTransactionId =
```

Comment:
I did that out of laziness / simplicity, but I think we don't really need the lock, since the value is only going to be changed by the current process. I could add a function to read the transactionNumber directly, adding a comment that it should only be used from the same session.